libbpf-rs/src/print.rs

```diff
@@ -24,8 +25,29 @@ impl From<libbpf_sys::libbpf_print_level> for PrintLevel {
 
 pub type PrintCallback = fn(PrintLevel, &str);
 
+/// Mimic the default print functionality of libbpf. This way if the user calls `set_print` when no
+/// previous callback had been set and then tries to restore it, it will appear to work correctly.
+fn default_callback(lvl: PrintLevel, msg: &str) {
+    if lvl == PrintLevel::Debug {
+        return;
+    }
+
+    let _ = io::stderr().write(msg.as_bytes());
+}
+
+// not allowed to use function pointers in const functions, so it needs to be a macro
+macro_rules! to_ptr {
+    ($x:expr) => {
+        unsafe { std::mem::transmute::<PrintCallback, *mut ()>($x) }
+    };
+}
+
+fn from_ptr(ptr: *mut ()) -> PrintCallback {
+    unsafe { std::mem::transmute::<*mut (), PrintCallback>(ptr) }
+}
+
 // There is no AtomicFnPtr. This is a workaround.
```

Comment:
Do you think using the [atomic crate](https://github.com/Amanieu/atomic-rs) for the AtomicFnPtr works as a viable alternative as well? It seems like other rust users are using the atomic crate for atomic function ptrs per the conversation in https://github.com/rust-lang/rfcs/issues/2481.

Reply:
I'll let the maintainers weigh in on this. It seemed unnecessary to add a dependency for a single use in an obscure little function, but I'm happy to do so.

Note that the use of atomics here is just a way to have global mutable state without locking or having questionable soundness. We can't actually say it's thread safe, because the underlying implementation of `libbpf_set_print` is not thread safe.